{
 "cells": [
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "from IPython.display import HTML\n",
    "\n",
    "HTML('''<script>\n",
    "code_show=true; \n",
    "function code_toggle() {\n",
    " if (code_show){\n",
    " $('div.input').hide();\n",
    " } else {\n",
    " $('div.input').show();\n",
    " }\n",
    " code_show = !code_show\n",
    "} \n",
    "$( document ).ready(code_toggle);\n",
    "</script>\n",
    "<form action=\"javascript:code_toggle()\"><input type=\"submit\" value=\"Click here to toggle on/off the raw code.\"></form>''')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "2019-02-04: INSeq Klebsiella Report\n",
    "===================================\n",
    "\n",
    "## Purpose\n",
    "\n",
    "* This notebook will contain an overview of the analysis done on INSeq data from a colonization experiment of *Klebsiella Pneumonia*\n",
    "\n",
    "### Computational Experiments:\n",
    "\n",
    "1. Pyinseq - gave a table with hits per gene in every sample, this was used in analysis\n",
    "2. DESeq Statistical Analysis - did statistical test taking into consideration the behavior of count data produced from high-throughput sequencing experiments.\n",
    "\n",
    "### Import summary gene table data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "    <div class=\"bk-root\">\n",
       "        <a href=\"https://bokeh.pydata.org\" target=\"_blank\" class=\"bk-logo bk-logo-small bk-logo-notebook\"></a>\n",
       "        <span id=\"1001\">Loading BokehJS ...</span>\n",
       "    </div>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/javascript": [
       "\n",
       "(function(root) {\n",
       "  function now() {\n",
       "    return new Date();\n",
       "  }\n",
       "\n",
       "  var force = true;\n",
       "\n",
       "  if (typeof (root._bokeh_onload_callbacks) === \"undefined\" || force === true) {\n",
       "    root._bokeh_onload_callbacks = [];\n",
       "    root._bokeh_is_loading = undefined;\n",
       "  }\n",
       "\n",
       "  var JS_MIME_TYPE = 'application/javascript';\n",
       "  var HTML_MIME_TYPE = 'text/html';\n",
       "  var EXEC_MIME_TYPE = 'application/vnd.bokehjs_exec.v0+json';\n",
       "  var CLASS_NAME = 'output_bokeh rendered_html';\n",
       "\n",
       "  /**\n",
       "   * Render data to the DOM node\n",
       "   */\n",
       "  function render(props, node) {\n",
       "    var script = document.createElement(\"script\");\n",
       "    node.appendChild(script);\n",
       "  }\n",
       "\n",
       "  /**\n",
       "   * Handle when an output is cleared or removed\n",
       "   */\n",
       "  function handleClearOutput(event, handle) {\n",
       "    var cell = handle.cell;\n",
       "\n",
       "    var id = cell.output_area._bokeh_element_id;\n",
       "    var server_id = cell.output_area._bokeh_server_id;\n",
       "    // Clean up Bokeh references\n",
       "    if (id != null && id in Bokeh.index) {\n",
       "      Bokeh.index[id].model.document.clear();\n",
       "      delete Bokeh.index[id];\n",
       "    }\n",
       "\n",
       "    if (server_id !== undefined) {\n",
       "      // Clean up Bokeh references\n",
       "      var cmd = \"from bokeh.io.state import curstate; print(curstate().uuid_to_server['\" + server_id + \"'].get_sessions()[0].document.roots[0]._id)\";\n",
       "      cell.notebook.kernel.execute(cmd, {\n",
       "        iopub: {\n",
       "          output: function(msg) {\n",
       "            var id = msg.content.text.trim();\n",
       "            if (id in Bokeh.index) {\n",
       "              Bokeh.index[id].model.document.clear();\n",
       "              delete Bokeh.index[id];\n",
       "            }\n",
       "          }\n",
       "        }\n",
       "      });\n",
       "      // Destroy server and session\n",
       "      var cmd = \"import bokeh.io.notebook as ion; ion.destroy_server('\" + server_id + \"')\";\n",
       "      cell.notebook.kernel.execute(cmd);\n",
       "    }\n",
       "  }\n",
       "\n",
       "  /**\n",
       "   * Handle when a new output is added\n",
       "   */\n",
       "  function handleAddOutput(event, handle) {\n",
       "    var output_area = handle.output_area;\n",
       "    var output = handle.output;\n",
       "\n",
       "    // limit handleAddOutput to display_data with EXEC_MIME_TYPE content only\n",
       "    if ((output.output_type != \"display_data\") || (!output.data.hasOwnProperty(EXEC_MIME_TYPE))) {\n",
       "      return\n",
       "    }\n",
       "\n",
       "    var toinsert = output_area.element.find(\".\" + CLASS_NAME.split(' ')[0]);\n",
       "\n",
       "    if (output.metadata[EXEC_MIME_TYPE][\"id\"] !== undefined) {\n",
       "      toinsert[toinsert.length - 1].firstChild.textContent = output.data[JS_MIME_TYPE];\n",
       "      // store reference to embed id on output_area\n",
       "      output_area._bokeh_element_id = output.metadata[EXEC_MIME_TYPE][\"id\"];\n",
       "    }\n",
       "    if (output.metadata[EXEC_MIME_TYPE][\"server_id\"] !== undefined) {\n",
       "      var bk_div = document.createElement(\"div\");\n",
       "      bk_div.innerHTML = output.data[HTML_MIME_TYPE];\n",
       "      var script_attrs = bk_div.children[0].attributes;\n",
       "      for (var i = 0; i < script_attrs.length; i++) {\n",
       "        toinsert[toinsert.length - 1].firstChild.setAttribute(script_attrs[i].name, script_attrs[i].value);\n",
       "      }\n",
       "      // store reference to server id on output_area\n",
       "      output_area._bokeh_server_id = output.metadata[EXEC_MIME_TYPE][\"server_id\"];\n",
       "    }\n",
       "  }\n",
       "\n",
       "  function register_renderer(events, OutputArea) {\n",
       "\n",
       "    function append_mime(data, metadata, element) {\n",
       "      // create a DOM node to render to\n",
       "      var toinsert = this.create_output_subarea(\n",
       "        metadata,\n",
       "        CLASS_NAME,\n",
       "        EXEC_MIME_TYPE\n",
       "      );\n",
       "      this.keyboard_manager.register_events(toinsert);\n",
       "      // Render to node\n",
       "      var props = {data: data, metadata: metadata[EXEC_MIME_TYPE]};\n",
       "      render(props, toinsert[toinsert.length - 1]);\n",
       "      element.append(toinsert);\n",
       "      return toinsert\n",
       "    }\n",
       "\n",
       "    /* Handle when an output is cleared or removed */\n",
       "    events.on('clear_output.CodeCell', handleClearOutput);\n",
       "    events.on('delete.Cell', handleClearOutput);\n",
       "\n",
       "    /* Handle when a new output is added */\n",
       "    events.on('output_added.OutputArea', handleAddOutput);\n",
       "\n",
       "    /**\n",
       "     * Register the mime type and append_mime function with output_area\n",
       "     */\n",
       "    OutputArea.prototype.register_mime_type(EXEC_MIME_TYPE, append_mime, {\n",
       "      /* Is output safe? */\n",
       "      safe: true,\n",
       "      /* Index of renderer in `output_area.display_order` */\n",
       "      index: 0\n",
       "    });\n",
       "  }\n",
       "\n",
       "  // register the mime type if in Jupyter Notebook environment and previously unregistered\n",
       "  if (root.Jupyter !== undefined) {\n",
       "    var events = require('base/js/events');\n",
       "    var OutputArea = require('notebook/js/outputarea').OutputArea;\n",
       "\n",
       "    if (OutputArea.prototype.mime_types().indexOf(EXEC_MIME_TYPE) == -1) {\n",
       "      register_renderer(events, OutputArea);\n",
       "    }\n",
       "  }\n",
       "\n",
       "  \n",
       "  if (typeof (root._bokeh_timeout) === \"undefined\" || force === true) {\n",
       "    root._bokeh_timeout = Date.now() + 5000;\n",
       "    root._bokeh_failed_load = false;\n",
       "  }\n",
       "\n",
       "  var NB_LOAD_WARNING = {'data': {'text/html':\n",
       "     \"<div style='background-color: #fdd'>\\n\"+\n",
       "     \"<p>\\n\"+\n",
       "     \"BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \\n\"+\n",
       "     \"may be due to a slow or bad network connection. Possible fixes:\\n\"+\n",
       "     \"</p>\\n\"+\n",
       "     \"<ul>\\n\"+\n",
       "     \"<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\\n\"+\n",
       "     \"<li>use INLINE resources instead, as so:</li>\\n\"+\n",
       "     \"</ul>\\n\"+\n",
       "     \"<code>\\n\"+\n",
       "     \"from bokeh.resources import INLINE\\n\"+\n",
       "     \"output_notebook(resources=INLINE)\\n\"+\n",
       "     \"</code>\\n\"+\n",
       "     \"</div>\"}};\n",
       "\n",
       "  function display_loaded() {\n",
       "    var el = document.getElementById(\"1001\");\n",
       "    if (el != null) {\n",
       "      el.textContent = \"BokehJS is loading...\";\n",
       "    }\n",
       "    if (root.Bokeh !== undefined) {\n",
       "      if (el != null) {\n",
       "        el.textContent = \"BokehJS \" + root.Bokeh.version + \" successfully loaded.\";\n",
       "      }\n",
       "    } else if (Date.now() < root._bokeh_timeout) {\n",
       "      setTimeout(display_loaded, 100)\n",
       "    }\n",
       "  }\n",
       "\n",
       "\n",
       "  function run_callbacks() {\n",
       "    try {\n",
       "      root._bokeh_onload_callbacks.forEach(function(callback) { callback() });\n",
       "    }\n",
       "    finally {\n",
       "      delete root._bokeh_onload_callbacks\n",
       "    }\n",
       "    console.info(\"Bokeh: all callbacks have finished\");\n",
       "  }\n",
       "\n",
       "  function load_libs(js_urls, callback) {\n",
       "    root._bokeh_onload_callbacks.push(callback);\n",
       "    if (root._bokeh_is_loading > 0) {\n",
       "      console.log(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n",
       "      return null;\n",
       "    }\n",
       "    if (js_urls == null || js_urls.length === 0) {\n",
       "      run_callbacks();\n",
       "      return null;\n",
       "    }\n",
       "    console.log(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n",
       "    root._bokeh_is_loading = js_urls.length;\n",
       "    for (var i = 0; i < js_urls.length; i++) {\n",
       "      var url = js_urls[i];\n",
       "      var s = document.createElement('script');\n",
       "      s.src = url;\n",
       "      s.async = false;\n",
       "      s.onreadystatechange = s.onload = function() {\n",
       "        root._bokeh_is_loading--;\n",
       "        if (root._bokeh_is_loading === 0) {\n",
       "          console.log(\"Bokeh: all BokehJS libraries loaded\");\n",
       "          run_callbacks()\n",
       "        }\n",
       "      };\n",
       "      s.onerror = function() {\n",
       "        console.warn(\"failed to load library \" + url);\n",
       "      };\n",
       "      console.log(\"Bokeh: injecting script tag for BokehJS library: \", url);\n",
       "      document.getElementsByTagName(\"head\")[0].appendChild(s);\n",
       "    }\n",
       "  };var element = document.getElementById(\"1001\");\n",
       "  if (element == null) {\n",
       "    console.log(\"Bokeh: ERROR: autoload.js configured with elementid '1001' but no matching script tag was found. \")\n",
       "    return false;\n",
       "  }\n",
       "\n",
       "  var js_urls = [\"https://cdn.pydata.org/bokeh/release/bokeh-1.0.2.min.js\", \"https://cdn.pydata.org/bokeh/release/bokeh-widgets-1.0.2.min.js\", \"https://cdn.pydata.org/bokeh/release/bokeh-tables-1.0.2.min.js\", \"https://cdn.pydata.org/bokeh/release/bokeh-gl-1.0.2.min.js\"];\n",
       "\n",
       "  var inline_js = [\n",
       "    function(Bokeh) {\n",
       "      Bokeh.set_log_level(\"info\");\n",
       "    },\n",
       "    \n",
       "    function(Bokeh) {\n",
       "      \n",
       "    },\n",
       "    function(Bokeh) {\n",
       "      console.log(\"Bokeh: injecting CSS: https://cdn.pydata.org/bokeh/release/bokeh-1.0.2.min.css\");\n",
       "      Bokeh.embed.inject_css(\"https://cdn.pydata.org/bokeh/release/bokeh-1.0.2.min.css\");\n",
       "      console.log(\"Bokeh: injecting CSS: https://cdn.pydata.org/bokeh/release/bokeh-widgets-1.0.2.min.css\");\n",
       "      Bokeh.embed.inject_css(\"https://cdn.pydata.org/bokeh/release/bokeh-widgets-1.0.2.min.css\");\n",
       "      console.log(\"Bokeh: injecting CSS: https://cdn.pydata.org/bokeh/release/bokeh-tables-1.0.2.min.css\");\n",
       "      Bokeh.embed.inject_css(\"https://cdn.pydata.org/bokeh/release/bokeh-tables-1.0.2.min.css\");\n",
       "    }\n",
       "  ];\n",
       "\n",
       "  function run_inline_js() {\n",
       "    \n",
       "    if ((root.Bokeh !== undefined) || (force === true)) {\n",
       "      for (var i = 0; i < inline_js.length; i++) {\n",
       "        inline_js[i].call(root, root.Bokeh);\n",
       "      }if (force === true) {\n",
       "        display_loaded();\n",
       "      }} else if (Date.now() < root._bokeh_timeout) {\n",
       "      setTimeout(run_inline_js, 100);\n",
       "    } else if (!root._bokeh_failed_load) {\n",
       "      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n",
       "      root._bokeh_failed_load = true;\n",
       "    } else if (force !== true) {\n",
       "      var cell = $(document.getElementById(\"1001\")).parents('.cell').data().cell;\n",
       "      cell.output_area.append_execute_result(NB_LOAD_WARNING)\n",
       "    }\n",
       "\n",
       "  }\n",
       "\n",
       "  if (root._bokeh_is_loading === 0) {\n",
       "    console.log(\"Bokeh: BokehJS loaded, going straight to plotting\");\n",
       "    run_inline_js();\n",
       "  } else {\n",
       "    load_libs(js_urls, function() {\n",
       "      console.log(\"Bokeh: BokehJS plotting callback run at\", now());\n",
       "      run_inline_js();\n",
       "    });\n",
       "  }\n",
       "}(window));"
      ],
      "application/vnd.bokehjs_load.v0+json": "\n(function(root) {\n  function now() {\n    return new Date();\n  }\n\n  var force = true;\n\n  if (typeof (root._bokeh_onload_callbacks) === \"undefined\" || force === true) {\n    root._bokeh_onload_callbacks = [];\n    root._bokeh_is_loading = undefined;\n  }\n\n  \n\n  \n  if (typeof (root._bokeh_timeout) === \"undefined\" || force === true) {\n    root._bokeh_timeout = Date.now() + 5000;\n    root._bokeh_failed_load = false;\n  }\n\n  var NB_LOAD_WARNING = {'data': {'text/html':\n     \"<div style='background-color: #fdd'>\\n\"+\n     \"<p>\\n\"+\n     \"BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \\n\"+\n     \"may be due to a slow or bad network connection. Possible fixes:\\n\"+\n     \"</p>\\n\"+\n     \"<ul>\\n\"+\n     \"<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\\n\"+\n     \"<li>use INLINE resources instead, as so:</li>\\n\"+\n     \"</ul>\\n\"+\n     \"<code>\\n\"+\n     \"from bokeh.resources import INLINE\\n\"+\n     \"output_notebook(resources=INLINE)\\n\"+\n     \"</code>\\n\"+\n     \"</div>\"}};\n\n  function display_loaded() {\n    var el = document.getElementById(\"1001\");\n    if (el != null) {\n      el.textContent = \"BokehJS is loading...\";\n    }\n    if (root.Bokeh !== undefined) {\n      if (el != null) {\n        el.textContent = \"BokehJS \" + root.Bokeh.version + \" successfully loaded.\";\n      }\n    } else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(display_loaded, 100)\n    }\n  }\n\n\n  function run_callbacks() {\n    try {\n      root._bokeh_onload_callbacks.forEach(function(callback) { callback() });\n    }\n    finally {\n      delete root._bokeh_onload_callbacks\n    }\n    console.info(\"Bokeh: all callbacks have finished\");\n  }\n\n  function load_libs(js_urls, callback) {\n    root._bokeh_onload_callbacks.push(callback);\n    if (root._bokeh_is_loading > 0) {\n      console.log(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n      return null;\n    }\n    if (js_urls == null || js_urls.length === 0) {\n      run_callbacks();\n      return null;\n    }\n    console.log(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n    root._bokeh_is_loading = js_urls.length;\n    for (var i = 0; i < js_urls.length; i++) {\n      var url = js_urls[i];\n      var s = document.createElement('script');\n      s.src = url;\n      s.async = false;\n      s.onreadystatechange = s.onload = function() {\n        root._bokeh_is_loading--;\n        if (root._bokeh_is_loading === 0) {\n          console.log(\"Bokeh: all BokehJS libraries loaded\");\n          run_callbacks()\n        }\n      };\n      s.onerror = function() {\n        console.warn(\"failed to load library \" + url);\n      };\n      console.log(\"Bokeh: injecting script tag for BokehJS library: \", url);\n      document.getElementsByTagName(\"head\")[0].appendChild(s);\n    }\n  };var element = document.getElementById(\"1001\");\n  if (element == null) {\n    console.log(\"Bokeh: ERROR: autoload.js configured with elementid '1001' but no matching script tag was found. \")\n    return false;\n  }\n\n  var js_urls = [\"https://cdn.pydata.org/bokeh/release/bokeh-1.0.2.min.js\", \"https://cdn.pydata.org/bokeh/release/bokeh-widgets-1.0.2.min.js\", \"https://cdn.pydata.org/bokeh/release/bokeh-tables-1.0.2.min.js\", \"https://cdn.pydata.org/bokeh/release/bokeh-gl-1.0.2.min.js\"];\n\n  var inline_js = [\n    function(Bokeh) {\n      Bokeh.set_log_level(\"info\");\n    },\n    \n    function(Bokeh) {\n      \n    },\n    function(Bokeh) {\n      console.log(\"Bokeh: injecting CSS: https://cdn.pydata.org/bokeh/release/bokeh-1.0.2.min.css\");\n      Bokeh.embed.inject_css(\"https://cdn.pydata.org/bokeh/release/bokeh-1.0.2.min.css\");\n      console.log(\"Bokeh: injecting CSS: https://cdn.pydata.org/bokeh/release/bokeh-widgets-1.0.2.min.css\");\n      Bokeh.embed.inject_css(\"https://cdn.pydata.org/bokeh/release/bokeh-widgets-1.0.2.min.css\");\n      console.log(\"Bokeh: injecting CSS: https://cdn.pydata.org/bokeh/release/bokeh-tables-1.0.2.min.css\");\n      Bokeh.embed.inject_css(\"https://cdn.pydata.org/bokeh/release/bokeh-tables-1.0.2.min.css\");\n    }\n  ];\n\n  function run_inline_js() {\n    \n    if ((root.Bokeh !== undefined) || (force === true)) {\n      for (var i = 0; i < inline_js.length; i++) {\n        inline_js[i].call(root, root.Bokeh);\n      }if (force === true) {\n        display_loaded();\n      }} else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(run_inline_js, 100);\n    } else if (!root._bokeh_failed_load) {\n      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n      root._bokeh_failed_load = true;\n    } else if (force !== true) {\n      var cell = $(document.getElementById(\"1001\")).parents('.cell').data().cell;\n      cell.output_area.append_execute_result(NB_LOAD_WARNING)\n    }\n\n  }\n\n  if (root._bokeh_is_loading === 0) {\n    console.log(\"Bokeh: BokehJS loaded, going straight to plotting\");\n    run_inline_js();\n  } else {\n    load_libs(js_urls, function() {\n      console.log(\"Bokeh: BokehJS plotting callback run at\", now());\n      run_inline_js();\n    });\n  }\n}(window));"
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import matplotlib as mpl\n",
    "from bokeh.io import output_notebook, show, push_notebook\n",
    "from bokeh.plotting import figure, output_file\n",
    "from bokeh.models import *\n",
    "from bokeh.layouts import column, row, gridplot\n",
    "\n",
    "import ipywidgets as widgets\n",
    "from ipywidgets import interact, interact_manual, widgets\n",
    "output_notebook()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Summary gene table contains the output results from the pyinseq run. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Contig</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Length</th>\n",
       "      <th>PID</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Synonym</th>\n",
       "      <th>Code</th>\n",
       "      <th>COG</th>\n",
       "      <th>Product</th>\n",
       "      <th>ket/xyl_E1004_01</th>\n",
       "      <th>ket/xyl_E1004_03</th>\n",
       "      <th>ket/xyl_E1004_04</th>\n",
       "      <th>propofol_E1004_05</th>\n",
       "      <th>propofol_E1004_06</th>\n",
       "      <th>propofol_E1004_08</th>\n",
       "      <th>input_E1004_13</th>\n",
       "      <th>input_E1004_14</th>\n",
       "      <th>input_E1004_15</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP009208</td>\n",
       "      <td>15</td>\n",
       "      <td>227</td>\n",
       "      <td>+</td>\n",
       "      <td>213</td>\n",
       "      <td>AIK78632.1</td>\n",
       "      <td>-</td>\n",
       "      <td>VK055_0001</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>lambda phage tail tape-measure family protein</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Contig  Start  End Strand  Length         PID Gene     Synonym Code COG  \\\n",
       "0  CP009208     15  227      +     213  AIK78632.1    -  VK055_0001    -   -   \n",
       "\n",
       "                                         Product  ket/xyl_E1004_01  \\\n",
       "0  lambda phage tail tape-measure family protein               0.0   \n",
       "\n",
       "   ket/xyl_E1004_03  ket/xyl_E1004_04  propofol_E1004_05  propofol_E1004_06  \\\n",
       "0               0.0               0.0                0.0                0.0   \n",
       "\n",
       "   propofol_E1004_08  input_E1004_13  input_E1004_14  input_E1004_15  \n",
       "0                0.0             0.0             0.0             0.0  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "summary_df = pd.read_csv(\"2019-01-31_E1004_renamed.txt\", sep='\\t')\n",
    "summary_df.head(1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Sample correlation was performed using **spearman correlation\n",
    "** Less sensitive to outliers\n",
    "\n",
    "* As expected,\n",
    "\n",
    "<img src=\"2019-01-31_E1004_spearman_heatmap.png\" width=800>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### And a T50 Test\n",
    "Counts how many hits made up for than 50% of all hits in the sample. Is used to show if there is a bottleneck in the INSeq experiment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: 'experiments/2018-10-26_pyinseq-runs/klebsiella-disruption_09/log.txt'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-3-9aba75c87dcc>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0;32mwith\u001b[0m \u001b[0mopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"experiments/2018-10-26_pyinseq-runs/klebsiella-disruption_09/log.txt\"\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mlog\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m     \u001b[0mnfifty_d\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m{\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m     \u001b[0;32mfor\u001b[0m \u001b[0mline\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mlog\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0;34m\"N50\"\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mline\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msplit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\" \"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m             \u001b[0mline\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mline\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msplit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\" \"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: 'experiments/2018-10-26_pyinseq-runs/klebsiella-disruption_09/log.txt'"
     ]
    }
   ],
   "source": [
    "with open(\"experiments/2018-10-26_pyinseq-runs/klebsiella-disruption_09/log.txt\") as log:\n",
    "    nfifty_d = {}\n",
    "    for line in log:\n",
    "        if \"N50\" in line.split(\" \"):\n",
    "            line = line.split(\" \")\n",
    "            if len(line) == 10:\n",
    "                pass\n",
    "            else:\n",
    "                sample = line[10].replace(\":\",\"\")\n",
    "                n50 = line[11].replace(\"\\n\", \"\")\n",
    "                nfifty_d[sample] = int(n50)\n",
    "\n",
    "fig_T50, ax_T50 = plt.subplots()\n",
    "fig_T50.set_size_inches(10,8)\n",
    "\n",
    "# Set x labels\n",
    "ax_T50.set_xticks(range(len(nfifty_d)))\n",
    "ax_T50.set_xticklabels(list(nfifty_d.keys()))\n",
    "\n",
    "ax_T50.bar(range(len(nfifty_d)), list(nfifty_d.values()), align='center')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Conclusion\n",
    "* No significant bottleneck was observed"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### DESeq statistical analysis with Fold-Change and pvalue calculations\n",
    "* After samples were processed by Pyinseq and hits were mapped to correspondent gene, the count table was submitted to the DESeq pipeline. Even though DESeq was orginally meant for studying differentially expressed genes by using RNA-Seq, the statistical tools it uses can be easily implemented for INSeq experiments since the data behaves the same way. This way we can normalize the data to avoid any bias by low-count genes and errors in the sequencing process. \n",
    "\n",
    "It outputs a p-adjusted value which uses Benjamini-Hochberg procedure to recude false true discoveries (Type-1 Errors). Also includes log2FC using input as the denominator of the ratio."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "deseq_df = pd.read_csv(\"2019-01-31_deseq_table_results.txt\", sep='\\t')\n",
    "columns = list(deseq_df.columns[2:])\n",
    "deseq_df.set_index(\"Synonym\").sort_values(\"Ketamine log10FC\").head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Previously selected genes\n",
    "* Previous genes were selected based on the fold-change regarding the input count values. However, after running both Mann-Whitney test and a Wald T-Test, from DESeq, many were shown to be statistically non-significant."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "previous_genes = [gene.replace(\"\\n\",\"\") for gene in open(\"genesOfInterest.txt\")]\n",
    "previous_genes = deseq_df.set_index(\"Synonym\").loc[previous_genes]\n",
    "\n",
    "fig_ella = plt.figure(figsize=(12,10), constrained_layout=True)\n",
    "grid_ella = mpl.gridspec.GridSpec(3,4, figure=fig_ella)\n",
    "\n",
    "\n",
    "r = 0\n",
    "c = 0\n",
    "for i in range(len(previous_genes)):\n",
    "    ax = fig_ella.add_subplot(grid_ella[r,c])\n",
    "    if c == 3:\n",
    "        r += 1\n",
    "        c = 0\n",
    "    else:\n",
    "        c += 1\n",
    "    \n",
    "    gene = {\n",
    "        \"Input\": list(previous_genes.iloc[i,16:]),\n",
    "        \"Propofol\": list(previous_genes.iloc[i,13:16]),\n",
    "        \"Ketamine\": list(previous_genes.iloc[i,10:13])\n",
    "    }\n",
    "    \n",
    "    ax.boxplot(list(gene.values()), labels=list(gene.keys()), whis=10)\n",
    "    ax.set_title(f\"{previous_genes.iloc[i,:].Gene} ({previous_genes.iloc[i,:].name})\")\n",
    "previous_genes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Volcano Plot\n",
    "* To visualize this, I have set up a volcano plot highlighting the previous genes of interest. There is a hover tool available which allows you to see metadata of that data point, try it. In addition, you can hide data points by toggling the legend name on the right legend panel."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make DataSources\n",
    "previous_source = ColumnDataSource(previous_genes)\n",
    "all_source = ColumnDataSource(deseq_df)\n",
    "\n",
    "# Trend Lines\n",
    "vline = Span(location=np.log10(1/4), dimension='height', line_color='black', line_width=1)\n",
    "hline = Span(location=0.05, dimension='width', line_color='black', line_width=1)\n",
    "\n",
    "# Initiate figures\n",
    "fig_propofol = figure(plot_height=400, plot_width=400, title=\"Propofol Volcano Plot\", \n",
    "            tools=\"crosshair,pan,reset, save,wheel_zoom, lasso_select, hover\",\n",
    "            x_range=[-5,5], y_range=[10**0,0.001], tooltips=[\n",
    "    (\"P-Value adj\", \"@{Propofol padj}\"),\n",
    "    (\"Synonym\",\"@Synonym\"),\n",
    "    (\"Gene_ID\", \"@Gene\"),\n",
    "    (\"Fold-Change\", \"@{Propofol log10FC}\")\n",
    "], y_axis_type=\"log\")\n",
    "\n",
    "fig_ketamine = figure(plot_height=400, plot_width=400, title=\"Ketamine Volcano Plot\", \n",
    "            tools=\"crosshair,pan,reset, save,wheel_zoom, lasso_select, hover\",\n",
    "            x_range=[-5,5], y_range=[10**0,0.001], tooltips=[\n",
    "    (\"P-Value adj\", \"@{Ketamine padj}\"),\n",
    "    (\"Synonym\",\"@Synonym\"),\n",
    "    (\"Gene_ID\", \"@Gene\"),\n",
    "    (\"Fold-Change\", \"@{Ketamine log10FC}\")\n",
    "], y_axis_type=\"log\")\n",
    "\n",
    "figures = [fig_ketamine,fig_propofol]\n",
    "\n",
    "fig_propofol.scatter(x=\"Propofol log10FC\", y=\"Propofol padj\", source=all_source, fill_color='gray', line_color=\"gray\", size=4, legend=\"All Genes\")\n",
    "fig_propofol.scatter(x=\"Propofol log10FC\", y=\"Propofol padj\", source=previous_source, fill_color=\"red\", line_color=\"red\", size=8, legend=\"Previous Genes\")\n",
    "\n",
    "fig_ketamine.scatter(x=\"Ketamine log10FC\", y=\"Ketamine padj\", source=all_source, fill_color='gray', line_color=\"gray\", size=4, legend=\"All Genes\")\n",
    "fig_ketamine.scatter(x=\"Ketamine log10FC\", y=\"Ketamine padj\", source=previous_source, fill_color=\"red\", line_color=\"red\", size=8, legend=\"Previous Genes\")\n",
    "\n",
    "for fig in figures:\n",
    "    fig.renderers.extend([vline,hline])\n",
    "    fig.legend.click_policy = \"hide\"\n",
    "    \n",
    "grid = gridplot([figures], toolbar_location=\"right\")\n",
    "\n",
    "show(grid)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Conclusion:\n",
    "Most of the genes chosen previously to study were determined to be **statistically insignificant**. These are:\n",
    "\n",
    "\\* 0.05 \n",
    "\n",
    "** 0.01\n",
    "\n",
    "| Synonym | Gene | Propofol | Ketamine |\n",
    "|---------|------|----------|----------|\n",
    "| VK055_2084 | - | NS | NS |\n",
    "| VK055_1930 | *fepC* | NS | NS |\n",
    "| VK055_1398 | - | NS | NS |\n",
    "| VK055_3638 | - | NS | NS |\n",
    "| VK055_0094 | *copA* | NS | NS |\n",
    "| VK055_4623 | *glnB* | NS | ** |\n",
    "| VK055_1993 | - | NS | *  |\n",
    "| VK055_4268 | - | NS | NS |\n",
    "| VK055_3462 | - | ** | NS |\n",
    "| VK055_3202 | *ilvC* | * | * |\n",
    "| VK055_3875 | *yrbC* | ** | ** |\n",
    "| \n",
    "\n",
    "\n",
    "\n",
    " "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Scatter plot showing candidate genes fold-change in significance in experimental conditions\n",
    "<img src=\"2019-01-10_f1_Draft3.png\" width=800>\n",
    "\n",
    "##### The genes marked by the legend were saved in tables\n",
    "\n",
    "* [Only down in Propofol](2019-02-01_propofol_genes.csv)\n",
    "* [Only down in Ketamine-Xylazine](2019-02-01_ketamine_genes.csv)\n",
    "* [Down both](2019-02-01_both_genes.csv)\n",
    "\n",
    "##### The scatter plot is interactive\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Data sources\n",
    "only_ket = ColumnDataSource(pd.read_csv(\"2019-02-01_ketamine_genes.csv\", sep=\"\\t\"))\n",
    "both_df = ColumnDataSource(pd.read_csv(\"2019-02-01_both_genes.csv\",sep='\\t'))\n",
    "only_prop = ColumnDataSource(pd.read_csv(\"2019-02-01_propofol_genes.csv\", sep=\"\\t\"))\n",
    "\n",
    "# Initiate figures\n",
    "fig_scatter = figure(plot_height=600, plot_width=600, title=\"Differentially Abundant Genes\", \n",
    "            tools=\"crosshair,pan,reset, save,wheel_zoom, lasso_select, hover\", tooltips=[\n",
    "    (\"Propofol P-Value adj\", \"@{Propofol padj}\"),\n",
    "    (\"Ketamine P-Value adj\", \"@{Ketamine padj}\"),\n",
    "    (\"Synonym\",\"@Synonym\"),\n",
    "    (\"Gene_ID\", \"@Gene\"),\n",
    "    (\"Propofol Fold-Change\", \"@{Propofol log10FC}\"),\n",
    "    (\"Ketamine Fold-Change\", \"@{Ketamine log10FC}\")],\n",
    "    x_axis_label='Propofol log10FC', y_axis_label='Ketamine log10FC')\n",
    "\n",
    "fig_scatter.scatter(x=\"Propofol log10FC\", y=\"Ketamine log10FC\", source=only_ket,\n",
    "                    fill_color=\"green\", line_color=\"green\", legend=\"Only in Propofol\")\n",
    "fig_scatter.scatter(x=\"Propofol log10FC\", y=\"Ketamine log10FC\", source=only_prop,\n",
    "                    fill_color=\"blue\", line_color=\"blue\", legend=\"Only in Ketamine/Xylazine\")\n",
    "fig_scatter.scatter(x=\"Propofol log10FC\", y=\"Ketamine log10FC\", source=both_df,\n",
    "                    fill_color=\"#e69f00\", line_color=\"#e69f00\", legend=\"Down in Both\")\n",
    "\n",
    "fig_scatter.legend.click_policy = \"hide\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "show(fig_scatter)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Possible Genes for studying further\n",
    "\n",
    "After revising genes that were statistically significant in this INSeq experiment, I searched for genes that were essential for the growth of Klebsiella and virulence.\n",
    "\n",
    "From this paper [Genome-Wide Identification of Klebsiella pneumoniae Fitness Genes during Lung Infection](https://mbio.asm.org/content/6/3/e00775-15.abstract), I selected a couple of genes and checked their Fold-Change under Propofol and Ketamine. These were:\n",
    "\n",
    "* rfaH - Transcriptional Activator, know for serum resistance\n",
    "* ilv operon - synthesis of branched amino-acids\n",
    "    *ilvC\n",
    "    *ilvD\n",
    "* VK055_4417 - MarR family protein\n",
    "* aroE - shikimate dehydrogenase required for aromatic amino acid synthesis\n",
    "\n",
    "From previously selected genes, this ones were significant in one or both conditions: \n",
    "* glnB - \n",
    "* VK055_1993\n",
    "* VK055_3462\n",
    "* yrbC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "synonym_int = [\"VK055_4417\", \"VK055_1993\", \"VK055_3462\"]\n",
    "gene_int = ['rfaH', 'ilvC', 'ilvD', 'glnB','yrbC', 'aroE']\n",
    "\n",
    "synonym_df = deseq_df.set_index(\"Synonym\")\n",
    "gene_df = deseq_df.set_index(\"Gene\")\n",
    "\n",
    "found_df = pd.DataFrame()\n",
    "for gene in gene_int:\n",
    "    f = gene_df.loc[gene]\n",
    "    found_df = found_df.append(f)\n",
    "found_df = found_df.reset_index().set_index(\"Synonym\")\n",
    "\n",
    "for syn in synonym_int:\n",
    "    f = synonym_df.loc[syn]\n",
    "    found_df = found_df.append(f)\n",
    "found_df = found_df.drop(\"Gene\", axis=1).rename(columns={\"index\":\"Gene\"})\n",
    "found_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Box plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig_genes = plt.figure(constrained_layout=True, figsize=(10,8))\n",
    "grid_gene = mpl.gridspec.GridSpec(3,3, figure=fig_genes)\n",
    "r,c = 0,0\n",
    "for i,row in found_df.iterrows():\n",
    "\n",
    "    ax = fig_genes.add_subplot(grid_gene[r,c])\n",
    "    if c == 2:\n",
    "        r += 1\n",
    "        c = 0\n",
    "    else:\n",
    "        c += 1\n",
    "    gene = {\n",
    "        \"Input\": list(row.iloc[10:13]),\n",
    "        \"Propofol\": list(row.iloc[16:]),\n",
    "        \"Ketamine\": list(row.iloc[13:16])\n",
    "    }\n",
    "    \n",
    "    ax.boxplot(list(gene.values()), labels=list(gene.keys()), whis=10)\n",
    "    if str(row.Gene) == \"nan\":\n",
    "        ax.set_title(f\"- ({row.name})\")\n",
    "    else:\n",
    "        ax.set_title(f\"{row.Gene} ({row.name})\")\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Select Genes Tool\n",
    "* Use the buttons below to filter genes below the desired Fold-Change (default=0.25) and pvalue (default=0.05)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "@interact\n",
    "def filter_df(condition=['Ketamine', 'Propofol', 'Both'], Fold_Change=widgets.FloatSlider(min=0,max=10,step=0.1, value=1/4) ,\n",
    "              pvalue=widgets.FloatSlider(min=0,max=0.1,step=0.01, value=0.05)):\n",
    "    \n",
    "    if condition == \"Both\":\n",
    "        output = deseq_df[ (deseq_df[\"Propofol\" + ' log10FC'] < np.log10(Fold_Change)) & (deseq_df[\"Propofol\" + \" padj\"] < pvalue)\n",
    "                         & (deseq_df[\"Ketamine\" + ' log10FC'] < np.log10(Fold_Change)) & (deseq_df[\"Ketamine\" + \" padj\"] < pvalue)]\n",
    "    elif condition != \"Both\":\n",
    "        output = deseq_df[ (deseq_df[condition + ' log10FC'] < np.log10(Fold_Change)) & (deseq_df[condition + \" padj\"] < pvalue)]\n",
    "    \n",
    "    button = widgets.Button(description=\"Save genes\")\n",
    "    \n",
    "\n",
    "    text_box = widgets.Textarea(\n",
    "        value='',\n",
    "        placeholder='Filename for saving genes',\n",
    "        description='Output name:',\n",
    "        disabled=False\n",
    "        )\n",
    "\n",
    "    display(button, text_box)\n",
    "    def save_genes(b, text_box=text_box):\n",
    "        output.to_csv(text_box.get_interact_value() + \".csv\", sep='\\t')\n",
    "        return\n",
    "    \n",
    "    button.on_click(save_genes)\n",
    "    return output\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
